js/modules/types/message.js

```diff
@@ -1,17 +1,165 @@
+// step. This will allow us to retroactively upgrade existing messages. As we
+// add more upgrade steps, we could design a pipeline that does this
+// incrementally, e.g. from version 0 / unknown -> 1, 1 --> 2, etc., similar to
+// how we do database migrations:
+exports.CURRENT_SCHEMA_VERSION = 2;
+
+
 // Public API
 exports.GROUP = GROUP;
 exports.PRIVATE = PRIVATE;
 
+// Placeholder until we have stronger preconditions:
+exports.isValid = () =>
+  true;
+
 // Schema
-// Message -> Promise Message
+exports.initializeSchemaVersion = (message) => {
+  const isInitialized = SchemaVersion.isValid(message.schemaVersion) &&
+    message.schemaVersion >= 1;
+  if (isInitialized) {
+    return message;
+  }
+
+  const numAttachments = Array.isArray(message.attachments)
+    ? message.attachments.length
+    : 0;
+  const hasAttachments = numAttachments > 0;
+  if (!hasAttachments) {
+    return Object.assign(
+      {},
+      message,
+      { schemaVersion: INITIAL_SCHEMA_VERSION }
+    );
+  }
+
+  // All attachments should have the same schema version, so we just pick
+  // the first one:
+  const firstAttachment = message.attachments[0];
+  const inheritedSchemaVersion = SchemaVersion.isValid(firstAttachment.schemaVersion)
+    ? firstAttachment.schemaVersion
+    : INITIAL_SCHEMA_VERSION;
+  const messageWithInitialSchema = Object.assign(
+    {},
+    message,
+    {
+      schemaVersion: inheritedSchemaVersion,
+      attachments: message.attachments.map(Attachment.removeSchemaVersion),
+    }
+  );
+
+  return messageWithInitialSchema;
+};
+
+// Middleware
+// type UpgradeStep = Message -> Promise Message
+
+// SchemaVersion -> UpgradeStep -> UpgradeStep
+exports._withSchemaVersion = (schemaVersion, upgrade) => {
+  if (!SchemaVersion.isValid(schemaVersion)) {
+    throw new TypeError('`schemaVersion` is invalid');
+  }
+  if (!isFunction(upgrade)) {
+    throw new TypeError('`upgrade` must be a function');
+  }
+
+  return async (message) => {
+    if (!exports.isValid(message)) {
+      console.log('Message._withSchemaVersion: Invalid input message:', message);
+      return message;
+    }
+
+    const isAlreadyUpgraded = message.schemaVersion >= schemaVersion;
+    if (isAlreadyUpgraded) {
+      return message;
+    }
+
+    const expectedVersion = schemaVersion - 1;
+    const hasExpectedVersion = message.schemaVersion === expectedVersion;
+    if (!hasExpectedVersion) {
+      console.log(
+        'WARNING: Message._withSchemaVersion: Unexpected version:',
+        `Expected message to have version ${expectedVersion},`,
+        `but got ${message.schemaVersion}.`,
+        message
+      );
+      return message;
+    }
+
+    let upgradedMessage;
+    try {
+      upgradedMessage = await upgrade(message);
+    } catch (error) {
+      console.log(
+        'Message._withSchemaVersion: error:',
+        Errors.toLogFormat(error)
+      );
+      return message;
```

Comment:
Interesting. If an error happens in migration v1, we'll get a log entry, but migration v2 will proceed. If there's no migration v2, then the message will be left still needing migration v1. Since the behavior in these two situations is different, maybe it makes sense to update to schema v1 even if migration v1 failed? The alternative is to stop processing entirely if we run into an error. We'd have to find out about the error and release an update to that problematic migration. It might be better to fail forward, and later realize that a given migration didn't always work.

What do you think?

Reply:
Good observation! As the test below confirms, we are indeed aborting the entire migration chain as soon as one step fails. The way I did that is ensuring that all migrations are consecutive, i.e. migration 3 has to follow 2 which has to follow 1, etc. using this code: https://github.com/signalapp/Signal-Desktop/blob/55b735027db2e26cd1e1a8c830dee8c8f6802d4c/js/modules/types/message.js#L98-L108 If it’s not, the data is passed through unchanged since function changes don’t have early abort unless we throw an error which I’d rather not if not necessary.

I believe this is critical as subsequent migrations should (and often must) be able to rely on information based on a given schema version. For example, if migration 3 moves attachments to disk, migration 4, 5, 6, etc. need to know that and work using `attachment.path` instead of `attachment.data`. If we start getting user debug logs with schema migration warnings, we’ll be able to investigate any problems. Let me know if you have any questions about this 😄